### ISSUE TITLE
clip_grad_norm_ silently passes when not finite

### ISSUE THREAD

**Issue by boeddeker** [pytorch/pytorch#46849]
## 🐛 Bug

The function `clip_grad_norm_` ignores non-finite values.
Suggestion: Raise an Exception.

## To Reproduce

```python
import torch
p = torch.ones(100, requires_grad=True, dtype=torch.float32)

print(torch.__version__, '\n')

for J in [
    p,
    p * 1e20,
    (p * torch.tensor(float('inf')))
]:
    J = J.sum()

    J.backward()
    print('max(abs(grad))\t', torch.max(torch.abs(p.grad)))
    print('grad_norm\t', torch.nn.utils.clip_grad_norm_(p, 1))
    print('max(abs(grad))\t', torch.max(torch.abs(p.grad)))
    print()
    p.grad = None
```
this yields on my notebook:
```
1.8.0.dev20201022+cpu 

max(abs(grad))	 tensor(1.)
grad_norm	 tensor(10.)
max(abs(grad))	 tensor(0.1000)

max(abs(grad))	 tensor(1.0000e+20)
grad_norm	 tensor(inf)
max(abs(grad))	 tensor(0.)

max(abs(grad))	 tensor(inf)
grad_norm	 tensor(inf)
max(abs(grad))	 tensor(nan)
```

## Expected behavior

Raise an Exception to report, that `grad_norm` calculation was not successfull and 
**don't change** the parameters.

Or alternatively, provide a non inplace version of `clip_grad_norm_`.
This allows the user to identify non finite values and provide a better Exception msg, 
e.g.: Tell which parameter caused the problem.
Currently, the function changes all values to non-interpretable values (either 0 or NaN).

## Environment

 - PyTorch Version (e.g., 1.0): 1.8.0.dev20201022+cpu 
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.6
 - CUDA/cuDNN version: None
 - GPU models and configuration: None
 - Any other relevant information: None

## Additional context

I opened this PR to open a discussion.
When someone tells me the desired behavior, I can make a PR.

cc @ezyang @gchanan @zou3519 @bdhirsh @heitorschueroff @albanD @mruberry

**Comment by ezyang** [comment]
> Or alternatively, provide a non inplace version of clip_grad_norm_.

Not sure about the other bits, but we should definitely do this.

**Comment by ezyang** [comment]
Raising an exception is not so easy when we're on CUDA.

**Comment by ezyang** [comment]
We briefly discussed this in our triage meeting but we weren't sure what the correct API was; in particular, raising an exception is not something we want to do because it would imply a CUDA sync, which this method doesn't currently do.

It might be easier for us to say what to do if you could describe to use what your intended remediation for this situation would be!

**Comment by boeddeker** [comment]
Ok, I try to explain my motivation.
In our framework around pytorch ([padertorch](https://github.com/fgnt/padertorch/blob/62ae19fd1ffe9056b49a3d6a4e32667082420326/padertorch/train/trainer.py#L704-L744)) we do some
cheap error checks to save the user some hours for the bug search.
So we check the loss if it is finite and report this to the user.

Sadly, it turns out, that it is more likely that a few gradients are not finite.
A check if all gradients are finite is expensive.
In most (or maybe all) experiments we use gradient clipping and report the norm to tensorboard.
So this value will be already calculated and checking if a scalar is finite is cheap (except for the cuda sync, but for us, it is ok).
So we check the norm and if it is not finite, we raise an Exception and save some training states for debugging.

When you then inspect the saved gradients, you only see zeros or NaNs.
So the gradients don't help to debug the trained NN (i.e. find the problematic layer).

One idea, when you are against an Exception (which implies a cuda sync) would be to split the `clip_grad_norm_` function, e.g.:
```python
def clip_grad_norm_(parameters, max_norm, ...):  # torch.nn.utils.clip_grad_norm_
    total_norm = grad_norm(parameters, ...)
    if norm > 1:
        scale_grad(parameters, max_norm / (total_norm + 1e-6))
    return norm
```
Then we could add a sync point between the two operations and only have to copy a few lines of code from pytorch.

An alternative would be to introduce a flag in the signature of `clip_grad_norm_` and raise only an exception when the flag is True.

**Comment by ezyang** [comment]
Thanks, this is very helpful.

For reference, here is the implementation of `clip_grad_norm_` right now:

```
    if isinstance(parameters, torch.Tensor):
        parameters = [parameters]
    parameters = list(filter(lambda p: p.grad is not None, parameters))
    max_norm = float(max_norm)
    norm_type = float(norm_type)
    if len(parameters) == 0:
        return torch.tensor(0.)
    device = parameters[0].grad.device
    if norm_type == inf:
        total_norm = max(p.grad.detach().abs().max().to(device) for p in parameters)
    else:
        total_norm = torch.norm(torch.stack([torch.norm(p.grad.detach(), norm_type).to(device) for p in parameters]), norm_type)
    clip_coef = max_norm / (total_norm + 1e-6)
    if clip_coef < 1:
        for p in parameters:
            p.grad.detach().mul_(clip_coef.to(p.grad.device))
    return total_norm
```

I notice that we do not actually do mutations until

**Comment by kurtamohler** [comment]
@ezyang  @mruberry, I can pick up this issue. But it sounds like there's still a question about whether raising an error is always the right thing to do. Should we just add a new argument to toggle old/new behavior like Ed suggests, just in case someone does not want it to error out? If so, I assume the argument should default to throwing an error, and the caller can elect to turn it off.

**Comment by ezyang** [comment]
Adding the argument seems cost free, so might as well do it.

**Comment by mruberry** [comment]
> @ezyang @mruberry, I can pick up this issue. But it sounds like there's still a question about whether raising an error is always the right thing to do. Should we just add a new argument to toggle old/new behavior like Ed suggests, just in case someone does not want it to error out? If so, I assume the argument should default to throwing an error, and the caller can elect to turn it off.

cc @jbschlosser

**Comment by jbschlosser** [comment]
> There is a mild BC concern if we start raising errors when there are infinite max_norm when there isn't otherwise. Whether or not it's OK to do this depends on whether or not there is any situation where silently turning the gradients into zero is helpful. If you always should be told when this situation occurs, we should just add the error check. Otherwise, we can add a new keyword argument to toggle between old and new behavior

I could see the behavior of silently turning inf-norm gradients into zero to be advantageous for some people. It allows training to continue with the gradient update effects of the current minibatch essentially "nulled out". Since the point of using `clip_grad_norm_` in general is to increase training stability (by avoiding exploding gradients), it could be annoying to have the `clip_grad_norm_` calculation be the cause of training being halted.

That said, I do think for safety's sake that it should be opt-in to avoid throwing an error, as there may well be cases where silently converting inf-norm gradients to zero is not the right thing to do.

Interestingly, TensorFlow's [`clip_by_global_norm`](https://www.tensorflow.org/api_docs/python/tf/clip_by_global_norm) chooses to set gradients to NaN for the inf-norm case to signal that an error has occurred. I personally prefer the more explicit behavior of throwing an error.

> If so, I assume the argument should default to throwing an error, and the caller can elect to turn it off.

I agree with @kurtamohler that this is the way to go. It's BC-breaking but more conservative by default.

**Comment by kurtamohler** [comment]
I've submitted a PR. I named the argument `error_if_nonfinite` for now. Other naming suggestions are more than welcome
